openedx_webhooks/github_views.py

```diff
@@ -46,29 +49,34 @@ def pull_request():
         return "PONG"
 
     pr = event["pull_request"]
+    PR_NUMBER = pr['number']
```

Comment:
I thought about this while swimming :)

# Why C has const

C provides for pointers to variables (and C++ has references). This means that variables can be affected by code far from their declaration.  I can pass a reference to a variable into a function, and the function can operate on the variable. The const declaration provides a way to prevent modification at a distance.

# Why JavaScript has const

I don't know why JavaScript has const. A five-minute search didn't turn up a good answer, just a lot of people pointing out that const is different than immutability. I suspect const is useful to the compiler or JIT engines, since the declaration provides information about the future that can guide their work.

# Why Python doesn't have const

A few reasons:

1. Unlike C, Python has no way to make a reference to a name, so there's no risk that a function will suprisingly change a variable passed into it.

2. Unlike JavaScript, Python has no declarations that could take a const modifier, so there's no way to provide guidance to the compiler or JIT.

3. Because Python only has naming conventions, and no declarations enforced by the language, we are relying on the name to guide the developer in their work.  This is very different than languages with a true const declaration that can be enforced in whatever way the language does.


# Semantic problems with "dynamically valued constants"

It seems like `PR_NUMBER = pr['number']` is stating that we are never going to change PR_NUMBER after this point.  Suppose I realize later that the value I got is a string, and I'd really find an integer more convenient.  Am I allowed to write this?

```python 
PR_NUMBER = pr['number']
PR_NUMBER = int(PR_NUMBER)
```

Seems like that would be wrong, since I am reassigning that name.  So can I write it like this?

```python
PR_NUMBER = int(pr['number'])
```

Now I've only assigned it once, so it's a constant. But that's a very subtle distinction between those two snippets, some might even say a trivial distinction.  Do we really want to be switching between lower-case and upper-case based on the number of assignment statements? It's an uninteresting consideration.

A constant is a thing whose value never changes.  "Dynamically-valued constants" is literally an oxymoron.

In the code in question, why was PR_NUMBER a "constant", but not other names?  By my reading of the code, we have eight other names with only a single assignment statement (event, pr, repo, action, ignored_actions, pr_activity, status_url, resp).   Why aren't they constants?

And then we get to the thornier issue of `result`, which is assigned in two different places, but only one of them will be executed in any given run of the function. Is that a constant?

Lastly, suppose we never reassign one of the parameters to a function? Should the parameter be named with all upper-case?  That would leak internal details to the callers of the function, and that detail could change as the function is further developed.

Overall, it seems to me that the number of times you assign to a name is an uninteresting detail, unless it can help the compiler or JIT.  Naming it with all upper-case letters in Python won't help the compiler or JIT.


# Why does Python have constants at all?

OK, so after all that, why use upper-case constants in Python at all? They aren't enforced, and you can still get into pedantic debates about multiple assignments as a way to construct the value.

First, true constants, names which refer to values which are always the same, every time you refer to them, are very useful in understanding a program. That a value never changes is an interesting detail to understand about a value.

Second, module-level or class-level constants are names that will be used across large bodies of code, far from where they were created.  Communicating the const-ness with a name is helpful to understanding those distant values.  Names within a single function don't have that same understanding-at-a-distance issue.